Add GitLab support to publish queued builds

tlusk opened the pull request:
This adds in support for publishing queued and canceled builds for the GitLab publisher.

Because TeamCity does build pipeline optimization and queues builds before deciding they don't need to be rebuilt, there needed to be some logic in here to handle those cases and publish the appropriate completed status back to GitLab if a queued build was optimized out of the queue.

There are potentially some other edge cases around multiple builds occurring on the same commit at the same time, but we've been using these changes heavily at our company for at least a few months now and haven't seen any major issues.

tlusk commented on commit-status-publisher-server/src/main/java/jetbrains/buildServer/commitPublisher/gitlab/GitlabPublisher.java:
```diff
@@ -112,9 +138,25 @@ private void publish(@NotNull String commit, @NotNull String data, @NotNull Repo
+      String responseString = response.getContent();
+      if (!responseString.contains("Cannot transition status via :enqueue from :pending") &&
+          !responseString.contains("Cannot transition status via :enqueue from :running") &&
+          !responseString.contains("Cannot transition status via :run from :running")) {
```
GitLab throws a 400 error when moving between certain states, even though it actually updates the status on GitLab.

There's a closed ticket where the resolution was to basically ignore the status code:
https://gitlab.com/gitlab-org/gitlab-ce/issues/25807

casz commented:
Could this PR get some love @dtretyakov ? 😊

karanagai commented:
Hi @tlusk,

Thank you for submitting this pull request. We do recongise that in some use cases publishing "pending" status when a build is queued would be beneficial, yet there are number of issues with it. You are addressing one of them, namely pending statuses staying on commits forever after respective builds were optimised by TeamCity in favour of builds running on later commits, but there is more to that, e.g. a revision that is calculated (rather guessed) for a queued build may not be the same as one that is used when the build is running. Also the way you are trying to restore the previous status of a commit based on the first queued build on the same branch may not work correctly in some cases anyway.

For that matter we are not ready to release this as an official feature, so for now we have decided to integrate most of your code except for that logic of restoring status. We have also disabled publishing build queued and removed from queue statuses by default, but allow users to enable it using `teamcity.commitStatusPublisher.publishQueuedBuildStatus` parameter (set to `true`) as a feature toggle.

Please let us know if this is ok by you.

tlusk commented:
@karanagai,

Everything you said above makes sense, and it's very understandable to not expose the queued status as an official feature as of yet. I'm just glad to see some of this pull request get merged into the publisher.

That said, while I'm sure there's some edge cases with how I implemented restoring the commit status for optimized builds (this was no surprise, I expected as much), I will say it has worked without issue for the past half a year with out TC & GitLab setup.

The optimization logic is a very critical component when using GitLab merge requests because without it it's very easy to get a "pending" status stuck on a commit forever making the merge request impossible to be merged. 

So I've updated and rebased the changes in this branch to now just include the logic around restoring the status of optimized builds so that others may be able to find and use them if they run into the same business needs as us and decide to use the unofficial queued build support.

I hope to in the future see an official fix to the "pending" forever issue, but until then we'll continue to use our forked version and I'll continue to keep the branch up to date.

tlusk commented:
Oh one question @karanagai, would it be worth submitting the changes around the queue optimization in another pull request but with a hidden configuration flag around it similar to how the queued build publishing is being enabled right now?
